### PR TITLE
Adding bearer token auth for metrics service

### PR DESCRIPTION
**Pull request by sivadeepN**
Motivation:

there is  no bearer token support in [k8s-monitoring-helm](https://github.com/grafana/k8s-monitoring-helm/tree/9b84eb42fc4e78beed119ff00d215478d6fa5508) chart
Adding the following changes to pass bearer token

resolves https://github.com/grafana/k8s-monitoring-helm/issues/614

### CODE REVIEW

**Comment by sivadeepN** [comment]
@petewall small change for bearer token support, can you review this

**Comment by petewall** [comment]
I'd love to bring this in, but there are a few changes that I'd like
First, can you make it so this only works when authMode = "bearerToken", then put the actual token inside a `bearerToken` section.

Also include a `___Key` field that explains how to store and reference it inside of the secret.

For example:
```
externalServices:
  prometheus:
    authMode: bearerToken
    bearerToken:
      token: my-token
      tokenKey: bearerToken
```

**Comment by petewall** [comment]
Here's a good PR that shows the kind of change that you should reference: https://github.com/grafana/k8s-monitoring-helm/pull/678

**Comment by sivadeepN** [comment]
@petewall following this notion as suggested

externalServices:
  prometheus:
    authMode: bearerToken
    bearerToken:
      token: my-token
      tokenFile: bearerTokenFile

Although not using the key since grafana alloy accepts file path for bearer token.
[Config ref](https://grafana.com/docs/alloy/latest/reference/components/prometheus/prometheus.remote_write/#endpoint-block)

File: charts/k8s-monitoring/README.md [hunk 1]
```diff
@@ -262,7 +262,7 @@ The Prometheus and Loki services may be hosted on the same cluster, or remotely
 
 | Key | Type | Default | Description |
 |-----|------|---------|-------------|
-| externalServices.prometheus.authMode | string | `"basic"` | one of "none", "basic", "oauth2" |
+| externalServices.prometheus.authMode | string | `"basic"` | one of "none", "basic", "oauth2" "bearerToken"|
```
Comment (petewall) [inline charts/k8s-monitoring/README.md:265]:
```suggestion
| externalServices.prometheus.authMode | string | `"basic"` | one of "none", "basic", "oauth2", "bearerToken"|
```

File: charts/k8s-monitoring/values.yaml [hunk 2]
```diff
@@ -117,6 +117,13 @@ externalServices:
       # @section -- External Services (Prometheus)
       tokenURL: ""
 
+    
+    # Authenticate to Prometheus using OAuth2
+    bearerToken: 
+      # Configure the Prometheus Remote Write Bearer Token
+      token: ""
```
Comment (petewall) [inline charts/k8s-monitoring/values.yaml:124]:
```suggestion
      # @section -- External Services (Prometheus)
      token: ""
      
      # -- The key for the token property in the secret
      # @section -- External Services (Prometheus)
      tokenKey: "token"      
```

File: charts/k8s-monitoring/values.yaml [hunk 3]
```diff
@@ -117,6 +117,13 @@ externalServices:
...
+      # Configure the Prometheus Remote Write Bearer Token secret
```
Comment (petewall) [inline charts/k8s-monitoring/values.yaml:125]:
```suggestion
      # Configure the Prometheus Remote Write Bearer Token secret
      # @section -- External Services (Prometheus)
```

File: charts/k8s-monitoring/templates/alloy_config/_metrics_service_remote_write.alloy.txt [hunk 4]
```diff
@@ -12,6 +12,12 @@ prometheus.remote_write "metrics_service" {
       username = nonsensitive(remote.kubernetes.secret.metrics_service.data[{{ .basicAuth.usernameKey | quote }}])
       password = remote.kubernetes.secret.metrics_service.data[{{ .basicAuth.passwordKey | quote }}]
     }
+{{- else if eq .authMode "bearerToken" }}
+  {{- if .bearerToken.token }}
+    bearer_token = {{  .bearerToken.token | quote  }}
```
Comment (petewall) [inline charts/k8s-monitoring/templates/alloy_config/_metrics_service_remote_write.alloy.txt:17]: Check out how we're storing secrets so we can store the bearer token in there.

**Comment by sivadeepN** [comment]
@petewall Here made changes for storing token in secret using tokenKey

Screenshot 2024-08-17 at 9 22 57 AM

File: charts/k8s-monitoring/templates/alloy_config/_metrics_service_remote_write.alloy.txt [hunk 5]
```diff
@@ -12,6 +12,12 @@ prometheus.remote_write "metrics_service" {
       username = nonsensitive(remote.kubernetes.secret.metrics_service.data[{{ .basicAuth.usernameKey | quote }}])
       password = remote.kubernetes.secret.metrics_service.data[{{ .basicAuth.passwordKey | quote }}]
     }
+{{- else if eq .authMode "bearerToken" }}
+  {{- if .bearerToken.token }}
+    bearer_token = remote.kubernetes.secret.metrics_service.data[{{ .bearerToken.tokenKey | quote }}]
+  {{- else if .bearerToken.tokenFile }}
+    bearer_token_file = {{  .bearerToken.tokenFile | quote  }}
+  {{- end }}
```
Comment (petewall) [inline charts/k8s-monitoring/templates/alloy_config/_metrics_service_remote_write.alloy.txt:20]:
```suggestion
  {{-  if .bearerToken.tokenFile }}
    bearer_token_file = {{  .bearerToken.tokenFile | quote  }}
  {{- else }}
    bearer_token = remote.kubernetes.secret.metrics_service.data[{{ .bearerToken.tokenKey | quote }}]
  {{- end }}
```

If a user uses an external (pre-created) secret, the bearerToken wont be in the values, but we still want to fetch it from the secret. So, that also means we should check if they passed the file first.

File: charts/k8s-monitoring/schema-mods/enums-and-types.json [hunk 6]
```diff
@@ -13,7 +13,7 @@
           "properties": {
             "authMode": {
               "type": "string",
-              "enum": ["none", "basic", "oauth2"]
+              "enum": ["none", "basic", "oauth2", "bearerToken"]
```
Comment (petewall) [inline charts/k8s-monitoring/schema-mods/enums-and-types.json:16]: Excellent!

**Review by petewall — COMMENTED** [review]
Just one small change and I think this is ready!

File: charts/k8s-monitoring/templates/alloy_config/_metrics_service_remote_write.alloy.txt (hunk 5 above)
Comment (shivangi-975) [reply charts/k8s-monitoring/templates/alloy_config/_metrics_service_remote_write.alloy.txt]: added changes as directed

**Review by petewall — APPROVED** [review]

**Pull request merged**